File: exercism/elixir/sum-of-multiples/sum_of_multiples.exs
defmodule SumOfMultiples do
  @doc """
  Adds up all numbers from 1 to a given end number that are multiples of the
  factors provided.
  """
  @spec to(non_neg_integer, [non_neg_integer]) :: non_neg_integer
  def to(limit, factors) do
    1..(limit - 1)
    |> Enum.filter(&factor_of_any?(&1, factors))
    |> Enum.sum
  end

  defp factor?(num, divider) do
    rem(num, divider) == 0
  end

  defp factor_of_any?(num, dividers) do
    Enum.any?(dividers, &factor?(num, &1))
  end
end

File: exercism/scala/sum-of-multiples/src/main/scala/SumOfMultiples.scala
object SumOfMultiples {
  def sum(factors: Set[Int], limit: Int): Int =
    (1 to limit - 1).filter(isMultiple(factors)).sum

  def isMultiple(factors: Set[Int])(num: Int): Boolean =
    factors.exists(num % _ == 0)
}
